### PR TITLE
Static Kernel Code Checker

### PR DESCRIPTION
**Pull request by simonguozirui**
Kernel Static Checker - Pattern-based validation for GPU kernel code.
See `src/kernel_static_checker.py`

The goal is flag reward hacking patterns (both strictly prohibited and possible ones).
through statically examining the code.

Warning: This list is by no means complete and nor this is not a replacement for runtime checks.
We welcome feedback and contributions as community find new ways of hacks.

- Bypass hacks (PyTorch wrapping, try-except fallback, inheritance bypass)
- Disallow some high-level torch operations (depends on the settings)
- Backend implementation requirements, that CUDA or DSL features must be used

Usage:
    result = validate_kernel_static(code, backend="cuda")
    will return a tuple (valid, errors, warnings) 

Right now, all these checks are regex-matching-based.
In the future we can add AST-based checking (@palic) to make it more reliable, more runtime level checks, and LM-as-a-judge. 

Source of checks:
- Kevin [paper](https://arxiv.org/abs/2507.11948), especially the reward hacking part (Appendix F)
- Snippets of code from our other projects
- DeepReinforce's awesome [blogpost](https://deep-reinforce.com/defense_kernel_hack.html)

### CODE REVIEW

File: uv.lock
Comment (PaliC) [inline uv.lock:1]: As discussed remove this :P

More seriously just update .gitignore

File: src/kernelbench/unit_tests/test_static_checker.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
+"""
+Tests for kernel_static_checker.py
```
Comment (PaliC) [inline src/kernelbench/unit_tests/test_static_checker.py:2]: Would recommend using fixtures for this (this is a common strat for web dev / js testing)

I can give you some examples if you want

File: src/code_check_old.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+        return True
+
+    # Additional check: no custom kernel implementation at all
+    has_triton_kernel = "@triton.jit" in kernel_code or "@triton.autotune" in kernel_code
+    has_cuda_kernel = "load_inline" in kernel_code or "cpp_extension" in kernel_code
+    has_cute_kernel = "cute::" in kernel_code or "from cutlass" in kernel_code
+    has_tilelang = "@T.prim_func" in kernel_code or "tvm.build" in kernel_code
+
+    has_custom_implementation = any([
+        has_triton_kernel,
```
Comment (PaliC) [inline src/code_check_old.py]: for the python dsl. An idea would be to patch the decorators such that they do some inspection op / timing to validate it's executed / it takes up most of the time

File: src/code_check_old.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+    1. PyTorch wrapping: Using torch.nn or torch.nn.functional
+    2. Try-except fallback: Wrapping incomplete CUDA in exception handlers
+    3. Inheritance bypass: Using 'pass' to inherit from reference
+
+    Reference: arXiv:2507.11948 "Kevin: Multi-Turn RL for Generating CUDA Kernels"
+    """
+    # Kevin Rule 1: PyTorch wrapping detection
+    # Zero reward for kernels containing torch.nn or torch.nn.functional
+    pytorch_patterns = [
+        "torch.nn.functional",
```
Comment (PaliC) [inline src/code_check_old.py]: I would just overwrite these libraries to emit a useful error if you want to ban torch

File: src/kernelbench/kernel_static_checker.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    "torch.max_pool1d", "torch.max_pool2d", "torch.max_pool3d",
+    "torch.adaptive_avg_pool1d", "torch.adaptive_avg_pool2d", "torch.adaptive_avg_pool3d",
+    "torch.adaptive_max_pool1d", "torch.adaptive_max_pool2d", "torch.adaptive_max_pool3d",
+    # Activations
+    "torch.relu", "torch.hardtanh", "torch.elu", "torch.selu",
+    "torch.leaky_relu", "torch.gelu", "torch.softsign", "torch.softplus",
+    "torch.softmax", "torch.log_softmax", "torch.tanh", "torch.sigmoid",
+    "torch.hardsigmoid", "torch.silu", "torch.mish",
+    # Normalization
+    "torch.batch_norm", "torch.group_norm", "torch.layer_norm",
```
Comment (PaliC) [inline src/kernelbench/kernel_static_checker.py:124]: just overwrite these with an error I think it's better.

File: src/kernelbench/unit_tests/test_static_checker.py (hunk 1 above)
Comment (simonguozirui) [reply src/kernelbench/unit_tests/test_static_checker.py:2]: yeah i will clean up the test file with some adveserial examples and reusable features, it was a start

File: src/code_check_old.py (hunk 2 above)
Comment (simonguozirui) [reply src/code_check_old.py]: good idea, we can use that in the runtime-based validation later on.

File: src/code_check_old.py (hunk 3 above)
Comment (simonguozirui) [reply src/code_check_old.py]: i see what you mean by overwriting now, let's keep this a static checker for now and we can think about smarter runtime checks later

**Comment by alexzhang13** [comment]
These are cool, I didn't take a careful enough look to make precise comments but in general I like the idea of all these tests. Also make sure people have to option to use or not use them (especially for the L3+ kernels where it might be too hard to not use PyTorch ops at all).

File: src/kernelbench/unit_tests/test_precision.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+various patterns that indicate precision downgrading (FP32 -> FP16),
+which is a reward hacking pattern.
+
+Note: The current implementation only checks for FP32 -> FP16 downgrades,
+not BF16 downgrades.
+
+Run with pytest:
+    pytest src/kernelbench/unit_tests/test_precision.py -v
+    or
+    uv run pytest src/kernelbench/unit_tests/test_precision.py -v
```
Comment (simonguozirui) [inline src/kernelbench/unit_tests/test_precision.py:14]: @ethanboneh  this is a bit redundant and too extensive,  perhaps we can clean it up with the real preicison checks in result PR

**Comment by simonguozirui** [comment]
Thanks to @alexzhang13 and @PaliC for the thoughtful reviews and discussions!

The goal of this PR is to introduce an initial checker system. While I like @PaliC’s AST-based checker idea, this PR focuses on a regex-based approach, which already catches a large class of problematic patterns in practice. Many of these patterns are derived from observations during Kevin RL training, as well as issues highlighted in Jiwei’s recent [blog post](https://deep-reinforce.com/defense_kernel_hack.html).

One advantage of regex-based checking is that it allows us to efficiently scan and validate large numbers of already-generated kernels. Going forward, we can complement this with more dynamic checkers in the evaluation pipeline, and I also plan to explore adding an LM-as-a-judge checker.

This work is part of our ongoing effort to mitigate reward hacking, as discussed in #74, and we hope it can serve as a useful resource for the community. The current set of patterns is not meant to be exhaustive, and contributions of additional checks are very welcome.

Since this system is still experimental, we will merge it but it is gated behind a flag. Users can configure groups of patterns that are either disallowed or treated as warnings, depending on their use case. See the file for more details on our suggestions for use cases.


particularly a note on degree of how much pytorch computational ops should be allowed
```
# Since KernelBench problems uses PyTorch as a reference, there could be settigs where
# Model generated code
# 1. Replaces some (not all) ops with custom kernels, others are kept in Torch
# --> More practical from a performance perspective (ie. make better systems) as you want to use whatever makes the best system for your use case. 
# 2. All compuational ops must be replaced with custom kernels
# --> Could be helpful from an eval (model ability on transpile + optimization) / RL training perspective 
# Depends the setting you use, you can move the checks below (pytorch_wrap, torch_computation_ops) 
# from WARNING to STRICT

# --- PyTorch NN Module Wrapping ---
# Allows: nn.Module, nn.Parameter, nn.ParameterList, nn.ParameterDict, 
#         nn.ModuleList, nn.ModuleDict, nn.init (needed for model structure)
# Blocks: nn.Linear, nn.Conv2d, nn.ReLU, etc. (compute layers)
```
